src/main/java/io/airlift/slice/SliceUtf8.java

```diff
@@ -0,0 +1,810 @@
+            if (!WHITESPACE_CODE_POINTS[codePoint]) {
+                break;
+            }
+            position += lengthOfCodePoint(codePoint);
+        }
+        return position;
+    }
+
+    /**
+     * Removes all white space characters from the right side of the string.
+     * </p>
+     * Note: This method does not explicitly check for valid UTF-8, and may
+     * return incorrect results or throw an exception for invalid UTF-8.
+     */
+    public static Slice rightTrim(Slice utf8)
+    {
+        int position = lastNonWhitespacePosition(utf8, 0);
+        return utf8.slice(0, position);
+    }
+
+    private static int lastNonWhitespacePosition(Slice utf8, int minPosition)
+    {
+        int length = utf8.length();
+
+        int position = length;
+        while (minPosition < position) {
+            int codePoint = getCodePointBefore(utf8, position);
+            if (!WHITESPACE_CODE_POINTS[codePoint]) {
+                break;
+            }
+            position -= lengthOfCodePoint(codePoint);
+        }
+        return position;
+    }
+
+    /**
+     * Removes all white space characters from the left and right side of the string.
+     * </p>
+     * Note: This method does not explicitly check for valid UTF-8, and may
+     * return incorrect results or throw an exception for invalid UTF-8.
+     */
+    public static Slice trim(Slice utf8)
+    {
+        int start = firstNonWhitespacePosition(utf8);
+        int end = lastNonWhitespacePosition(utf8, start);
+        return utf8.slice(start, end - start);
+    }
+
+    public static Slice fixInvalidUtf8(Slice slice)
+    {
+        return fixInvalidUtf8(slice, REPLACEMENT_CODE_POINT);
+    }
+
+    public static Slice fixInvalidUtf8(Slice slice, int replacementCodePoint)
+    {
+        if (isAscii(slice)) {
+            return slice;
+        }
+
+        int replacementCodePointLength = lengthOfCodePoint(replacementCodePoint);
+
+        int length = slice.length();
+        Slice utf8 = Slices.allocate(length);
+
+        int dataPosition = 0;
+        int utf8Position = 0;
+        while (dataPosition < length) {
+            int codePoint = getCodePointAtSafe(slice, dataPosition);
+            int codePointLength;
+            if (codePoint >= 0) {
+                codePointLength = lengthOfCodePoint(codePoint);
+                dataPosition += codePointLength;
+            }
+            else {
+                // negative number carries the number of invalid bytes
+                dataPosition += (-codePoint);
+                codePoint = replacementCodePoint;
+                codePointLength = replacementCodePointLength;
+            }
+            utf8 = Slices.ensureSize(utf8, utf8Position + codePointLength);
+            utf8Position += setCodePointAt(codePoint, utf8, utf8Position);
+        }
+        return utf8.slice(0, utf8Position);
+    }
+
+    private static int getCodePointAtSafe(Slice utf8, int position)
```

Comment:
This function uses `getByte` for 1st and 2nd byte, and `getByteUnchecked` for 3rd byte and onwards. This inconsistency appears to be a bug.

* If `getByteUnchecked` is used on 1st byte, this function should be named "unchecked".
* Otherwise, this function need to do a bound check before reading 2nd-nth byte using `getByteUnchecked`. Otherwise, it need to be named "unchecked".